presto-native-execution/presto_cpp/main/tests/PrestoExchangeSourceTest.cpp

```diff
@@ -684,26 +685,36 @@ TEST_P(PrestoExchangeSourceTest, slowProducer) {
   ASSERT_EQ(stats.at("prestoExchangeSource.totalBytes"), totalBytes(pages));
 }
 
-TEST_P(PrestoExchangeSourceTest, slowProducerAndEarlyTerminatingConsumer) {
+DEBUG_ONLY_TEST_P(
+    PrestoExchangeSourceTest,
+    slowProducerAndEarlyTerminatingConsumer) {
   const bool useHttps = GetParam().useHttps;
-  std::atomic<bool> codePointHit{false};
+  std::atomic_bool codePointHit{false};
+  folly::EventCount closeWait;
+  std::atomic_bool allCloseCheckPassed{false};
   SCOPED_TESTVALUE_SET(
       "facebook::presto::PrestoExchangeSource::doRequest",
+      std::function<void(const PrestoExchangeSource*)>(
+          ([&](const auto* prestoExchangeSource) {
+            allCloseCheckPassed = true;
+          })));
+  SCOPED_TESTVALUE_SET(
+      "facebook::presto::PrestoExchangeSource::doRequest::thenValue",
       std::function<void(const PrestoExchangeSource*)>(
           ([&](const auto* prestoExchangeSource) { codePointHit = true; })));
-  auto producer = std::make_unique<Producer>();
 
+  auto producer = std::make_unique<Producer>();
   auto producerServer = createHttpServer(useHttps);
   producer->registerEndpoints(producerServer.get());
-
   test::HttpServerWrapper serverWrapper(std::move(producerServer));
   auto producerAddress = serverWrapper.start().get();
 
   auto queue = makeSingleSourceQueue();
   auto exchangeSource = makeExchangeSource(producerAddress, useHttps, 3, queue);
-
   requestNextPage(queue, exchangeSource);
 
+  closeWait.await([&]() { return allCloseCheckPassed.load(); });
```

Comment:
If the failure is codePointHit is not true, can we just wait for codePointHit is true?
And set it to true in
facebook::presto::PrestoExchangeSource::handleResponse in case that we have refactored the code? Thanks!

Reply:
We need to pre-kill the exchange source to recreate the race condition. The wait on notification before requestBuilder's callback is to recreate the race condition.

If we directly wait for codePointHit before killing then we set codePointHit to true when the exchange source is not yet closed, which defies the purpose of the test.